rlog/2023-09-27-gossipimprovements.mdx

```diff
@@ -0,0 +1,206 @@
+3) Node failure can result in partitions, making many segments unreachable. 
+4) Assuring a shortest-path tree-like structure requires a detailed view of the underlying (and continuously changing) network topology. 
+
+
+Solutions involve creating multiple random trees to add redundancy [2].
+Alternatives involve building an overlay mesh and forwarding messages through the multicast delivery tree (eager push). 
+
+Metadata is shared through the overlay links so that the nodes can ask for missing messages (lazy push or pull-based operation) through the overlay links. 
+New nodes are added from the overlay on node failure, but it requires non-faulty node selection.
+
+GossipSub uses eager push (through overlay mesh) and lazy push (through IWANT messages). 
+
+The mesh degree $D_{Low} \leq D \leq D_{High}$ is crucial in deciding message dissemination latency. 
+A smaller value for $D$ results in higher latency due to increased rounds, whereas a higher $D$ reduces latency on the cost of increased redundancy. 
+It is suggested that the average mesh degree should be $D_{avg} = \ln(N) + C$ for an optimal operation, where $N$ is the network size and C is a small constant. 
+
+Moreover, quicker shuffling of peers results in better performance in the presence of resource-constrained nodes and/or node failure [3]. 
+
+## GOAL2: Considering Heterogeneity In Overlay Design
+Random peering connections in P2P overlays represent a stochastic process. It is inherently difficult to precisely model the performance of such systems. 
+Most of the research on P2P networks provides simulation results assuming nodes with similar capabilities. 
+The aspect of dissimilar capabilities and resource-constrained nodes is less explored.
+
+It is discussed in GOAL1 that overlay mesh results in better performance if $D_{avg} = \ln(N) + C$. 
+Enforcing all the nodes to have approximately $\ln(N) + C$ peers makes resource-rich nodes under-utilized, while resource-constrained nodes are overloaded. 
+At the same time, connecting high-bandwidth nodes through a low-bandwidth node undermines the network's performance.
+Ideally, every node should contribute proportionally to the available resources.
+A better solution involves a two-phased operation:
+
+
+1. Every node computes its available bandwidth and selects a node degree $D_i$ proportional to its available bandwidth.
+Different bandwidth estimation approaches are suggested in literature [4,5]. 
+
+we suggest using a simple bandwidth estimation approach similar to the ones used by dynamic adaptive video streaming over HTTP (DASH) [6]. 
+To further conform to the suggested mesh-degree average $D_{avg}$, every node tries achieving this number within its neighborhood, resulting in an overall similar mesh-degree average. 
+
+2. From the available local view, every node tries connecting peers with the lowest latency until $D$ connections are made.
+We suggest referring to the peering solution discussed in GOAL5 to avoid partitioning.

```

Comment:
This generates a very interesting discussion. I believe this will create partitions (clusters), where the nearest pears communicate with each other (at very low latency). That is why, I suggested in Goal 5 that high-capacity peers (from each cluster) must create an additional overlay.